File: sdks/typescript/src/conformance.ts
/** Conformance corpus runner — executes `conformance/vectors/*.json`.
 * Mirrors `protocol/conformance.py`; the same vector files drive every
 * SDK's conformance suite. */

import { readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { canonicalString } from "./canon.js";
import type { JsonValue } from "./canon.js";
import { ReceiptError } from "./errors.js";
import {
  PortableReceipt,
  ReceiptEdge,
  ReceiptNode,
  verifyProofDocument,
  KeyError,
} from "./receipt.js";
import { sign, verifySignatures } from "./signatures.js";
import { verifyAny } from "./verify.js";

const TAG = "$rr";
const TAGS: Record<string, unknown> = {
  nan: NaN,
  "+inf": Infinity,
  "-inf": -Infinity,
  "-0": -0,
};

function untag(value: unknown): unknown {
  if (value !== null && typeof value === "object" && !Array.isArray(value)) {
    const o = value as Record<string, unknown>;
    if (Object.keys(o).length === 1 && TAG in o) {
      const tag = o[TAG] as string;
      if (tag in TAGS) return TAGS[tag];
      throw new Error(`unknown tag ${tag}`);
    }
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(o)) out[k] = untag(v);
    return out;
  }
  if (Array.isArray(value)) return value.map(untag);
  return value;
}

interface BuildSpec {
  subject: string;
  metadata?: Record<string, JsonValue>;
  nodes?: { id: string; kind: string; payload?: JsonValue; meta?: Record<string, JsonValue> }[];
  edges?: { from: string; to: string; rel: string }[];
  receipt_id?: string;
  produced_at?: string;
}

function build(spec: BuildSpec): PortableReceipt {
  const nodes = (spec.nodes ?? []).map(
    (n) => new ReceiptNode(n.id, n.kind, untag(n.payload) as JsonValue, n.meta),
  );
  const edges = (spec.edges ?? []).map((e) => new ReceiptEdge(e.from, e.to, e.rel));
  return new PortableReceipt({
    subject: spec.subject,
    metadata: spec.metadata ?? {},
    nodes,
    edges,
    receipt_id: spec.receipt_id ?? "rr-conformance",
    produced_at: spec.produced_at ?? "2026-01-01T00:00:00Z",
  });
}

type Result = [boolean, string];

function runCanon(vector: Record<string, unknown>): Result {
  const expect = vector.expect as Record<string, unknown>;
  try {
    const produced = canonicalString(untag(vector.input) as JsonValue);
    if ("error" in expect) return [false, `expected error ${expect.error}, encoded fine`];
    return [produced === expect.canonical, "canonical bytes differ"];
  } catch (exc) {
    if (exc instanceof ReceiptError && "error" in expect) {
      return [expect.error === exc.code, `error ${exc.code} vs ${expect.error}`];
    }
    return [false, `unexpected error ${exc instanceof ReceiptError ? exc.code : String(exc)}`];
  }
}

function runReceipt(vector: Record<string, unknown>): Result {
  const expect = vector.expect as Record<string, unknown>;
  let receipt: PortableReceipt;
  let envelope: Record<string, JsonValue>;
  try {
    receipt = build(vector.input as BuildSpec);
    envelope = receipt.committedEnvelope();
  } catch (exc) {
    if (exc instanceof ReceiptError && "error" in expect) {
      return [expect.error === exc.code, `error ${exc.code} vs ${expect.error}`];
    }
    return [false, `unexpected error ${exc instanceof ReceiptError ? `${exc.code}: ${exc.message}` : String(exc)}`];
  }
  if ("error" in expect) return [false, `expected error ${expect.error}, built fine`];
  const mismatches: string[] = [];
  if (JSON.stringify(sortObj(envelope.node_hashes)) !== JSON.stringify(sortObj(expect.node_hashes))) {
    mismatches.push("node_hashes");
  }
  if (JSON.stringify(envelope.edge_hashes) !== JSON.stringify(expect.edge_hashes)) {
    mismatches.push("edge_hashes");
  }
  if (envelope.merkle_root !== expect.merkle_root) mismatches.push("merkle_root");
  if ("receipt_hash" in expect && receipt.receiptHash() !== expect.receipt_hash) {
    mismatches.push("receipt_hash");
  }
  if ("canonical_envelope" in expect) {
    const produced = canonicalString(envelope);
    if (produced !== expect.canonical_envelope) mismatches.push("canonical_envelope");
  }
  return [mismatches.length === 0, `fields differ: ${mismatches.join(", ")}`];
}

function runDocument(vector: Record<string, unknown>): Result {
  const expect = vector.expect as Record<string, unknown>;
  const document = untag(vector.input) as Record<string, JsonValue>;
  const rep = verifyAny(document);
  if (rep.valid !== expect.valid) {
    return [false, `valid=${rep.valid} expected ${expect.valid} (${rep.errors})`];
  }
  for (const code of (expect.errors as string[]) ?? []) {
    if (!rep.errors.some((e) => e.includes(code))) {
      return [false, `missing expected error code ${code} in ${JSON.stringify(rep.errors)}`];
    }
  }
  return [true, ""];
}

function runProof(vector: Record<string, unknown>): Result {
  const expect = vector.expect as Record<string, unknown>;
  if ("proof_document" in vector) {
    const ok = verifyProofDocument(untag(vector.proof_document));
    return [ok === expect.verify, `verify=${ok} expected ${expect.verify}`];
  }
  const receipt = build(vector.input as BuildSpec);
  let proofDoc: ReturnType<PortableReceipt["proofFor"]>;
  if ("node" in expect) {
    try {
      proofDoc = receipt.proofFor(expect.node as string);
    } catch (exc) {
      if (exc instanceof KeyError && "error" in expect) return [true, ""];
      return [false, "node not found, no error expected"];
    }
  } else {
    proofDoc = receipt.proofForEdge(expect.edge_index as number);
  }
  if (proofDoc.leaf !== expect.leaf) return [false, "leaf differs"];
  if (JSON.stringify(proofDoc.proof) !== JSON.stringify(expect.proof)) {
    return [false, "proof siblings differ"];
  }
  if (proofDoc.merkle_root !== expect.merkle_root) return [false, "root differs"];
  return [verifyProofDocument(proofDoc), "generated proof does not verify"];
}

function runSignature(vector: Record<string, unknown>): Result {
  const expect = vector.expect as Record<string, unknown>;
  const committed = build(vector.input as BuildSpec).committedEnvelope();
  const spec = vector.sign as Record<string, string>;
  const sigObj = sign(committed, spec.private_key, {
    scope: spec.scope ?? "receipt",
    key_id: spec.key_id,
    signed_at: spec.signed_at,
  });
  if ("public_key" in expect && sigObj.public_key !== expect.public_key) {
    return [false, "public_key differs"];
  }
  if ("sig" in expect && sigObj.sig !== expect.sig) {
    return [false, "signature bytes differ"];
  }
  const target =
    "verify_against" in vector
      ? build(vector.verify_against as BuildSpec).committedEnvelope()
      : committed;
  const results = verifySignatures(target, [sigObj]);
  const ok = results[0].valid;
  return [ok === (expect.valid ?? true), `verify=${ok}`];
}

const RUNNERS: Record<string, (v: Record<string, unknown>) => Result> = {
  canon: runCanon,
  receipt: runReceipt,
  document: runDocument,
  proof: runProof,
  signature: runSignature,
};

function sortObj(o: unknown): unknown {
  if (o !== null && typeof o === "object" && !Array.isArray(o)) {
    const out: Record<string, unknown> = {};
    for (const k of Object.keys(o).sort()) out[k] = sortObj((o as Record<string, unknown>)[k]);
    return out;
  }
  return o;
}

export interface VectorResult {
  name: string;
  ok: boolean;
  detail: string;
}

export function runVector(vector: Record<string, unknown>, name: string): VectorResult {
  const runner = RUNNERS[vector.kind as string];
  try {
    const [ok, detail] = runner(vector);
    return { name, ok, detail };
  } catch (exc) {
    const expect = (vector.expect ?? {}) as Record<string, unknown>;
    if (exc instanceof ReceiptError) {
      const ok = expect.error === exc.code;
      return {
        name,
        ok,
        detail: `error ${exc.code}` + (ok ? "" : ` (wanted ${expect.error})`),
      };
    }
    return { name, ok: false, detail: `${(exc as Error).constructor.name}: ${(exc as Error).message}` };
  }
}

export interface CorpusReport {
  vectors: VectorResult[];
  total: number;
  passed: number;
  ok: boolean;
  implementation: string;
}

export function runCorpus(vectorDir: string): CorpusReport {
  const files = readdirSync(vectorDir)
    .filter((f) => f.endsWith(".json"))
    .sort();
  const entries = files.map((f) => {
    const vector = JSON.parse(readFileSync(join(vectorDir, f), "utf-8"));
    return runVector(vector, vector.name ?? f.replace(/\.json$/, ""));
  });
  const passed = entries.filter((e) => e.ok).length;
  return {
    vectors: entries,
    total: entries.length,
    passed,
    ok: passed === entries.length && entries.length > 0,
    implementation: "typescript",
  };
}

File: sdks/rust/src/conformance.rs
//! Conformance corpus runner — executes `conformance/vectors/*.json`.
//! Mirrors `protocol/conformance.py`; the same vector files drive every
//! SDK's conformance suite.

use crate::canon::canonical_string;
use crate::errors::{ReceiptError, E_NONCANONICAL};
use crate::receipt::{
    verify_proof_document, PortableReceipt, ReceiptEdge, ReceiptNode,
};
use crate::signatures::{sign, verify_signatures};
use crate::verify::verify_any;
use serde_json::{json, Map, Value};
use std::fs;
use std::path::Path;

const TAG: &str = "$rr";

/// Tagged inputs (`{"$rr": "…"}`) express values JSON literals can't.
/// serde_json::Value cannot hold non-finite floats, so nan/inf tags are
/// rejected here with `noncanonical_value` — the same code Python's
/// canonicalizer emits one stage later when it sees the real float.
fn untag(v: &Value) -> Result<Value, ReceiptError> {
    match v {
        Value::Object(o) if o.len() == 1 && o.contains_key(TAG) => match o[TAG].as_str() {
            Some("nan") | Some("+inf") | Some("-inf") => Err(ReceiptError::new(
                E_NONCANONICAL,
                "non-finite number is not portable",
            )),
            Some("-0") => Ok(json!(-0.0f64)),
            other => panic!("unknown tag {other:?}"),
        },
        Value::Object(o) => o
            .iter()
            .map(|(k, v)| untag(v).map(|u| (k.clone(), u)))
            .collect::<Result<Map<String, Value>, _>>()
            .map(Value::Object),
        Value::Array(a) => a
            .iter()
            .map(untag)
            .collect::<Result<Vec<Value>, _>>()
            .map(Value::Array),
        _ => Ok(v.clone()),
    }
}

fn build(spec: &Value) -> PortableReceipt {
    let nodes: Vec<ReceiptNode> = spec
        .get("nodes")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
        .iter()
        .map(|n| ReceiptNode {
            id: n["id"].as_str().unwrap_or_default().to_string(),
            kind: n["kind"].as_str().unwrap_or_default().to_string(),
            payload: untag(n.get("payload").unwrap_or(&Value::Null))
                .unwrap_or(Value::Null),
            meta: n.get("meta").cloned(),
        })
        .collect();
    let edges: Vec<ReceiptEdge> = spec
        .get("edges")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
        .iter()
        .map(|e| ReceiptEdge {
            src: e["from"].as_str().unwrap_or_default().to_string(),
            dst: e["to"].as_str().unwrap_or_default().to_string(),
            rel: e["rel"].as_str().unwrap_or_default().to_string(),
        })
        .collect();
    PortableReceipt {
        subject: spec["subject"].as_str().unwrap_or_default().to_string(),
        metadata: spec
            .get("metadata")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default(),
        nodes,
        edges,
        receipt_id: spec
            .get("receipt_id")
            .and_then(Value::as_str)
            .unwrap_or("rr-conformance")
            .to_string(),
        produced_at: spec
            .get("produced_at")
            .and_then(Value::as_str)
            .unwrap_or("2026-01-01T00:00:00Z")
            .to_string(),
        schema_version: crate::receipt::SCHEMA_VERSION.to_string(),
        signatures: Vec::new(),
    }
}

type RunResult = std::result::Result<(bool, String), ReceiptError>;

fn run_canon(vector: &Value) -> RunResult {
    let expect = &vector["expect"];
    match untag(&vector["input"]).and_then(|v| canonical_string(&v)) {
        Ok(produced) => {
            if expect.get("error").is_some() {
                return Ok((false, format!("expected error {}, encoded fine", expect["error"])));
            }
            Ok((produced == expect["canonical"].as_str().unwrap_or(""), "canonical bytes differ".into()))
        }
        Err(exc) => {
            if let Some(want) = expect.get("error").and_then(Value::as_str) {
                return Ok((want == exc.code, format!("error {} vs {want}", exc.code)));
            }
            Ok((false, format!("unexpected error {}: {}", exc.code, exc.message)))
        }
    }
}

fn run_receipt(vector: &Value) -> RunResult {
    let expect = &vector["expect"];
    let receipt = build(&vector["input"]);
    let envelope = match receipt.committed_envelope() {
        Ok(e) => e,
        Err(exc) => {
            if let Some(want) = expect.get("error").and_then(Value::as_str) {
                return Ok((want == exc.code, format!("error {} vs {want}", exc.code)));
            }
            return Ok((false, format!("unexpected error {}: {}", exc.code, exc.message)));
        }
    };
    if expect.get("error").is_some() {
        return Ok((false, format!("expected error {}, built fine", expect["error"])));
    }
    let mut mismatches = Vec::new();
    if envelope["node_hashes"] != expect["node_hashes"] {
        mismatches.push("node_hashes");
    }
    if envelope["edge_hashes"] != expect["edge_hashes"] {
        mismatches.push("edge_hashes");
    }
    if envelope["merkle_root"] != expect["merkle_root"] {
        mismatches.push("merkle_root");
    }
    if let Some(want) = expect.get("receipt_hash") {
        if receipt.receipt_hash().ok().as_deref() != want.as_str() {
            mismatches.push("receipt_hash");
        }
    }
    if let Some(want) = expect.get("canonical_envelope").and_then(Value::as_str) {
        if canonical_string(&envelope).unwrap_or_default() != want {
            mismatches.push("canonical_envelope");
        }
    }
    Ok((mismatches.is_empty(), format!("fields differ: {}", mismatches.join(", "))))
}

fn run_document(vector: &Value) -> RunResult {
    let expect = &vector["expect"];
    let document = match untag(&vector["input"]) {
        Ok(d) => d,
        Err(exc) => {
            let want = expect["valid"].as_bool().unwrap_or(false);
            return Ok((!want, format!("untag error {}", exc.code)));
        }
    };
    let report = verify_any(&document, None);
    if report.valid != expect["valid"].as_bool().unwrap_or(false) {
        return Ok((
            false,
            format!("valid={} expected {} ({:?})", report.valid, expect["valid"], report.errors),
        ));
    }
    if let Some(codes) = expect.get("errors").and_then(Value::as_array) {
        for code in codes {
            let code = code.as_str().unwrap_or("");
            if !report.errors.iter().any(|e| e.contains(code)) {
                return Ok((
                    false,
                    format!("missing expected error code {code} in {:?}", report.errors),
                ));
            }
        }
    }
    Ok((true, String::new()))
}

fn run_proof(vector: &Value) -> RunResult {
    let expect = &vector["expect"];
    if let Some(doc) = vector.get("proof_document") {
        let ok = untag(doc).map(|d| verify_proof_document(&d)).unwrap_or(false);
        return Ok((ok == expect["verify"].as_bool().unwrap_or(false), format!("verify={ok} expected {}", expect["verify"])));
    }
    let receipt = build(&vector["input"]);
    let proof_doc = if let Some(node) = expect.get("node") {
        match receipt.proof_for(node.as_str().unwrap_or("")) {
            Ok(p) => p,
            Err(_) => {
                return if expect.get("error").is_some() {
                    Ok((true, String::new()))
                } else {
                    Ok((false, "node not found, no error expected".into()))
                }
            }
        }
    } else {
        match receipt.proof_for_edge(expect["edge_index"].as_u64().unwrap_or(0) as usize) {
            Ok(p) => p,
            Err(e) => return Ok((false, format!("edge proof error: {e}"))),
        }
    };
    if proof_doc["leaf"] != expect["leaf"] {
        return Ok((false, "leaf differs".into()));
    }
    if proof_doc["proof"] != expect["proof"] {
        return Ok((false, "proof siblings differ".into()));
    }
    if proof_doc["merkle_root"] != expect["merkle_root"] {
        return Ok((false, "root differs".into()));
    }
    Ok((verify_proof_document(&proof_doc), "generated proof does not verify".into()))
}

fn run_signature(vector: &Value) -> RunResult {
    let expect = &vector["expect"];
    let committed = build(&vector["input"]).committed_envelope()?;
    let spec = &vector["sign"];
    let sig_obj = sign(
        &committed,
        spec["private_key"].as_str().unwrap_or(""),
        spec.get("scope").and_then(Value::as_str).unwrap_or("receipt"),
        spec.get("key_id").and_then(Value::as_str),
        spec.get("signed_at").and_then(Value::as_str),
        None,
    )?;
    if let Some(want) = expect.get("public_key") {
        if sig_obj["public_key"] != *want {
            return Ok((false, "public_key differs".into()));
        }
    }
    if let Some(want) = expect.get("sig") {
        if sig_obj["sig"] != *want {
            return Ok((false, "signature bytes differ".into()));
        }
    }
    let target = match vector.get("verify_against") {
        Some(v) => build(v).committed_envelope()?,
        None => committed,
    };
    let results = verify_signatures(&target, vec![sig_obj])?;
    let ok = results[0]["valid"].as_bool().unwrap_or(false);
    Ok((ok == expect.get("valid").and_then(Value::as_bool).unwrap_or(true), format!("verify={ok}")))
}

pub struct VectorResult {
    pub name: String,
    pub ok: bool,
    pub detail: String,
}

pub fn run_vector(vector: &Value, name: &str) -> VectorResult {
    let kind = vector["kind"].as_str().unwrap_or("");
    let result: RunResult = match kind {
        "canon" => run_canon(vector),
        "receipt" => run_receipt(vector),
        "document" => run_document(vector),
        "proof" => run_proof(vector),
        "signature" => run_signature(vector),
        _ => return VectorResult { name: name.into(), ok: false, detail: format!("unknown kind {kind}") },
    };
    match result {
        Ok((ok, detail)) => VectorResult { name: name.into(), ok, detail },
        Err(exc) => {
            let want = vector["expect"]["error"].as_str();
            let ok = want == Some(exc.code);
            VectorResult {
                name: name.into(),
                ok,
                detail: format!("error {}{}", exc.code, if ok { "".into() } else { format!(" (wanted {})", want.unwrap_or("?")) }),
            }
        }
    }
}

pub struct CorpusReport {
    pub vectors: Vec<VectorResult>,
    pub total: usize,
    pub passed: usize,
    pub ok: bool,
}

/// Replace every `\uXXXX` surrogate-range escape with `\ufffd` so
/// serde_json can parse a vector's `expect` block after the strict
/// parser has already classified the file. Never applied to input bytes.
fn sanitize_surrogate_escapes(text: &str) -> String {
    let bytes = text.as_bytes();
    let mut out = String::with_capacity(text.len());
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'\\'
            && i + 6 <= bytes.len()
            && bytes[i + 1] == b'u'
            && matches!(bytes[i + 2], b'd' | b'D')
        {
            if let Ok(cp) = u32::from_str_radix(&text[i + 2..i + 6], 16) {
                if (0xd800..=0xdfff).contains(&cp) {
                    out.push_str("\\ufffd");
                    i += 6;
                    continue;
                }
            }
        }
        let ch_len = utf8_char_len(bytes[i]);
        out.push_str(&text[i..i + ch_len]);
        i += ch_len;
    }
    out
}

fn utf8_char_len(first: u8) -> usize {
    match first {
        0x00..=0x7f => 1,
        0xc0..=0xdf => 2,
        0xe0..=0xef => 3,
        _ => 4,
    }
}

pub fn run_corpus(vector_dir: &Path) -> CorpusReport {
    let mut files: Vec<_> = fs::read_dir(vector_dir)
        .expect("vector dir")
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().is_some_and(|x| x == "json"))
        .collect();
    files.sort();
    let vectors: Vec<VectorResult> = files
        .iter()
        .map(|p| {
            let text = fs::read_to_string(p).expect("read vector");
            let name = p.file_stem().unwrap().to_str().unwrap().to_string();
            // The strict parser can itself reject a vector file (e.g. the
            // lone-surrogate input). When it does, the vector's expected
            // error code is the contract — same observable result as
            // Python rejecting one stage later at encode time.
            match crate::canon::parse_json_value(&text) {
                Ok(vector) => {
                    let name = vector
                        .get("name")
                        .and_then(Value::as_str)
                        .unwrap_or(&name)
                        .to_string();
                    run_vector(&vector, &name)
                }
                Err(exc) => {
                    // Recover the expect block from a surrogate-sanitized
                    // copy (the strict parser already produced its code).
                    let sanitized = sanitize_surrogate_escapes(&text);
                    let vector: Value = serde_json::from_str(&sanitized).unwrap_or(json!({}));
                    let want = vector["expect"]["error"].as_str();
                    let ok = want == Some(exc.code);
                    VectorResult {
                        name,
                        ok,
                        detail: format!(
                            "error {}{}",
                            exc.code,
                            if ok { String::new() } else { format!(" (wanted {})", want.unwrap_or("?")) }
                        ),
                    }
                }
            }
        })
        .collect();
    let passed = vectors.iter().filter(|v| v.ok).count();
    CorpusReport {
        total: vectors.len(),
        passed,
        ok: passed == vectors.len() && !vectors.is_empty(),
        vectors,
    }
}
